lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+use OCP\Share\IShare;
+use Throwable;
+use GuzzleHttp\Exception\ClientException;
+use GuzzleHttp\Exception\ConnectException;
+use GuzzleHttp\Exception\ServerException;
+use OCP\Http\Client\IClientService;
+use OCP\Files\IRootFolder;
+use Psr\Log\LoggerInterface;
+use OCP\Share\IManager as ShareManager;
+use function preg_replace;
+
+class ImageService {
+
+	/**
+	 * @var ShareManager
+	 */
+	private $shareManager;
+	/**
+	 * @var IRootFolder
+	 */
+	private $rootFolder;
+	/**
+	 * @var IClientService
+	 */
+	private $clientService;
+	/**
+	 * @var LoggerInterface
+	 */
+	private $logger;
+	/**
+	 * @var IPreview
+	 */
+	private $previewManager;
+
+	public function __construct(IRootFolder $rootFolder,
+								LoggerInterface $logger,
+								ShareManager $shareManager,
+								IPreview $previewManager,
+								IClientService $clientService) {
+		$this->rootFolder = $rootFolder;
+		$this->shareManager = $shareManager;
+		$this->clientService = $clientService;
+		$this->logger = $logger;
+		$this->previewManager = $previewManager;
+	}
+
+	/**
+	 * Get image content or preview from file id
+	 * @param int $documentId
+	 * @param string $imageFileName
+	 * @param string $userId
+	 * @return File|\OCP\Files\Node|ISimpleFile|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function getImage(int $documentId, string $imageFileName, string $userId) {
+		$textFile = $this->getTextFile($documentId, $userId);
+		return $textFile === null ? null : $this->getImagePreview($imageFileName, $textFile);
+	}
+
+	/**
+	 * Get image content or preview from file id in public context
+	 * @param int $documentId
+	 * @param string $imageFileName
+	 * @param string $shareToken
+	 * @return File|\OCP\Files\Node|ISimpleFile|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function getImagePublic(int $documentId, string $imageFileName, string $shareToken) {
+		$textFile = $this->getTextFilePublic($documentId, $shareToken);
+		return $textFile === null ? null : $this->getImagePreview($imageFileName, $textFile);
+	}
+
+	/**
+	 * @param string $imageFileName
+	 * @param File $textFile
+	 * @return File|\OCP\Files\Node|ISimpleFile|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getImagePreview(string $imageFileName, File $textFile) {
+		$attachmentFolder = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($attachmentFolder !== null) {
+			try {
+				$imageFile = $attachmentFolder->get($imageFileName);
+			} catch (NotFoundException $e) {
+				return null;
+			}
+			if ($imageFile instanceof File) {
+				if ($this->previewManager->isMimeSupported($imageFile->getMimeType())) {
+					return $this->previewManager->getPreview($imageFile, 1024, 1024);
+				} else {
```

Comment:
From PhpStorm: [EA] Child instructions can be extracted here (clearer intention, lower complexity numbers).

Reply:
I don't get this on my side. What does it mean? :grin:

Reply:
It basically means that you can remove the `else {` statement and instead put everything inside `else` **behind** the `if` statement.

It's about changing

```php
if {
    ...
    return x;
} else {
    do_something();
}
```

to

```php
if {
    ...
    return x;
}
do_something();
```

Reply:
See https://github.com/kalessil/phpinspectionsea/blob/master/docs/control-flow.md#redundant-else-keyword for details :blush:

I guess it's the setting in Editor -> Inspections -> PHP -> Control flow -> Unnecessary 'else' branch.

Reply:
:+1: thanks